### PR TITLE
When launching wsl, promote the starting directory to --cd

### LINKED ISSUE #592: Support linux paths for `startingDirectory` of WSL distros
* Your Windows build number:  `Microsoft Windows [Version 10.0.18362.86]`

* What you're doing and what's happening:  
Setting the starting directory for wsl in `profiles.json` doesn't make it point to the right directory.  
1. Take the config below :-  
```json
{
        "startingDirectory": "C:/Users/joshj/AppData/Local/Packages/CanonicalGroupLimited.Ubuntu16.04onWindows_79rhkp1fndgsc/LocalState/rootfs/home/tackyunicorn",
        "guid": "{a2785f8e-72c9-4550-b406-697388644902}",
        "name": "Ubuntu",
        "colorscheme": "MonokaiVivid",
        "historySize": 9001,
        "snapOnInput": true,
        "cursorColor": "#FFFFFF",
        "cursorHeight": 25,
        "cursorShape": "vintage",
        "commandline": "wsl.exe",
        "fontFace": "Inconsolata for Powerline",
        "fontSize": 10,
        "acrylicOpacity": 1,
        "useAcrylic": true,
        "closeOnExit": true,
        "padding": "10, 10, 10, 10",
        "icon": "ms-appdata:///roaming/ubuntu.png"
}
```  
This results in wsl starting up in the root folder  
issue1

2. In this config :-  
```json
{
        "startingDirectory": "/home/tackyunicorn",
        "guid": "{a2785f8e-72c9-4550-b406-697388644902}",
        "name": "Ubuntu",
        "colorscheme": "MonokaiVivid",
        "historySize": 9001,
        "snapOnInput": true,
        "cursorColor": "#FFFFFF",
        "cursorHeight": 25,
        "cursorShape": "vintage",
        "commandline": "wsl.exe",
        "fontFace": "Inconsolata for Powerline",
        "fontSize": 10,
        "acrylicOpacity": 1,
        "useAcrylic": true,
        "closeOnExit": true,
        "padding": "10, 10, 10, 10",
        "icon": "ms-appdata:///roaming/ubuntu.png"
}
```  
This results in wsl starting up in the Windows user folder  
issue2

* What's wrong / what should be happening instead:  
The `startingDirectory` profile setting is not being applied correctly on a wsl profile

### PR DESCRIPTION
**Pull request by DHowett**
This commit introduces a hack to ConptyConnection for launching WSL.
When we detect that WSL is being launched (either "wsl" or "wsl.exe",
unqialified or _specifically_ from the current OS's System32 directory),
we will promote the startingDirectory specified at launch time into a
commandline argument.

Why do we want to switch to `--cd`?
With the current design of ConptyConnection and WSL, there are some
significant limitations:
* `startingDirectory` cannot be a WSL path, which forces users to
  use weird tricks such as setting the starting directory to
  `\\wsl$\Distro\home\user`.
* WSL occasionally fails to launch in time to handle a `\\wsl$` path,
  which makes us spawn in a strange location (or no location at all).

(This fix will only address the second one until a WSL update is
released that adds support for `--cd $LINUX_PATH`.)

We will not do the promotion if any of the following are true:
* the commandline contains `--cd` already
* the commandline contains a bare `~`
   * This was a commonly-used workaround that forced wsl to start in the
     user's home directory. It conflicts with --cd.
* wsl is not spelled properly (`WSL` and `WSL.EXE` are unacceptable)
* an absolute path to wsl outside the system32 directory is provided

We chose the do this trick in the connection layer, the latest possible
point, because it captures the most use cases.

We could have done it earlier, but the options were quite limiting.
They are:

* Generate WSL profiles with startingDirectory set to the home folder
   * We can't do this because we do not know the user's home folder
     path.
* Generate WSL profiles with `--cd` in them.
   * This only works for unmodified profiles.
   * This only works for generated profiles.
   * Users cannot override the commandline without breaking it.
   * Users cannot specify a startingDirectory (!) since the one on the
     commandline wins.
* Set a flag on generated WSL profiles to request this trick
   * This only works for generated profiles. Users who create their own
     WSL profiles couldn't set startingDirectory and have it work the
     same.

Patching the commandline, hacky though it may be, seemed to be the most
compatible option. Eventually, we can even support `wt -d ~ wsl`!

## Validation Steps Performed

Manual validation for the following cases:

```c++
// MUST MANGLE
auto a01 = _tryMangleStartingDirectoryForWSL(LR"(wsl)", L"SENTINEL");
auto a02 = _tryMangleStartingDirectoryForWSL(LR"(wsl -d X)", L"SENTINEL");
auto a03 = _tryMangleStartingDirectoryForWSL(LR"(wsl -d X ~/bin/sh)", L"SENTINEL");
auto a04 = _tryMangleStartingDirectoryForWSL(LR"(wsl.exe)", L"SENTINEL");
auto a05 = _tryMangleStartingDirectoryForWSL(LR"(wsl.exe -d X)", L"SENTINEL");
auto a06 = _tryMangleStartingDirectoryForWSL(LR"(wsl.exe -d X ~/bin/sh)", L"SENTINEL");
auto a07 = _tryMangleStartingDirectoryForWSL(LR"("wsl")", L"SENTINEL");
auto a08 = _tryMangleStartingDirectoryForWSL(LR"("wsl.exe")", L"SENTINEL");
auto a09 = _tryMangleStartingDirectoryForWSL(LR"("wsl" -d X)", L"SENTINEL");
auto a10 = _tryMangleStartingDirectoryForWSL(LR"("wsl.exe" -d X)", L"SENTINEL");
auto a11 = _tryMangleStartingDirectoryForWSL(LR"("C:\Windows\system32\wsl.exe" -d X)", L"SENTINEL");
auto a12 = _tryMangleStartingDirectoryForWSL(LR"("C:\windows\system32\wsl" -d X)", L"SENTINEL");
auto a13 = _tryMangleStartingDirectoryForWSL(LR"(wsl ~/bin)", L"SENTINEL");

// MUST NOT MANGLE
auto a14 = _tryMangleStartingDirectoryForWSL(LR"("C:\wsl.exe" -d X)", L"SENTINEL");
auto a15 = _tryMangleStartingDirectoryForWSL(LR"(C:\wsl.exe)", L"SENTINEL");
auto a16 = _tryMangleStartingDirectoryForWSL(LR"(wsl --cd C:\)", L"SENTINEL");
auto a17 = _tryMangleStartingDirectoryForWSL(LR"(wsl ~)", L"SENTINEL");
auto a18 = _tryMangleStartingDirectoryForWSL(LR"(wsl ~ -d Ubuntu)", L"SENTINEL");
```

We don't have anywhere to put TerminalConnection unit tests :|

Closes #592.

### CODE REVIEW

**Comment by driver1998** [comment]
It sounds unlikely, but what'll happen if I have a different wsl.exe somewhere else in %PATH%?

**Comment by DHowett** [comment]
My official stance is, “it had better be compatible with wsl, because otherwise it should not be masquerading as wsl.”

**Review by miniksa — APPROVED** [review]
Good enough. Just get the typo.

File: src/cascadia/TerminalConnection/ConptyConnection.cpp [hunk 1]
```diff
@@ -56,6 +56,72 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
         return S_OK;
     }
 
+    // Function Descrption:
```
Comment (zadjii-msft) [inline src/cascadia/TerminalConnection/ConptyConnection.cpp:59]:
```suggestion
    // Function Description:
```

File: src/cascadia/TerminalConnection/ConptyConnection.cpp [hunk 2]
```diff
@@ -56,6 +56,72 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
...
+    // - Promotes a starting directory provided to a WSL invocation to a commandline argument.
+    //   This is necessary because WSL has some modicum of support for linux-side directories (!) which
+    //   CreateProcess never will.
+    static std::tuple<std::wstring, std::wstring> _tryMangleStartingDirectoryForWSL(std::wstring_view commandLine, std::wstring_view startingDirectory)
+    {
+        do
+        {
+            if (startingDirectory.size() > 0 && commandLine.size() >= 3)
+            { // "wsl" is three characters; this is a safe bet. no point in doing it if there's no starting directory though!
```
Comment (zadjii-msft) [inline src/cascadia/TerminalConnection/ConptyConnection.cpp:68]:
```suggestion
            { 
                // "wsl" is three characters; this is a safe bet. no point in doing it if there's no starting directory though!
```

File: src/cascadia/TerminalConnection/ConptyConnection.cpp [hunk 3]
```diff
@@ -56,6 +56,72 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
...
+
+                    const auto tilde{ arguments.find_first_of(L'~') };
+                    if (tilde != std::wstring_view::npos)
+                    {
+                        if (tilde + 1 == arguments.size() || til::at(arguments, tilde + 1) == L' ')
+                        {
+                            // We want to suppress --cd if they have added a bare ~ to their commandline (they conflict).
+                            break;
+                        }
+                        // Tilde followed by non-space should be okay (like, wsl -d Debian ~/blah.sh)
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalConnection/ConptyConnection.cpp:108]: Yea I'm mildly worried about someone doing some silly thing like `wsl -- bash -c vim ~ ; ping whatever.com` 

(that's made up but you get the point)

**Review by zadjii-msft — APPROVED** [review]
* this feels dangerous, but hey let's ship it to Preview and see what happens
* We should really _really_ have connection tests with this kind of code getting added. That of course means another project (😢), but at least it's a project that's not dependent on WinUI?

**Comment by zadjii-msft** [comment]
@DHowett fix the type and yeet this for the bug bash this afternoon

**Comment by DHowett** [comment]
@zadjii-msft thx for the reminder. this is building in selfhost-1.9 right now. I'll merge after the bash :)

**Comment by zadjii-msft** [comment]
@DHowett do we want to try this on for 1.11? maybe even behind velocity if we're nervous?

**Comment by DHowett** [comment]
We are *still* normalizing these somewhere.

```
1:23:41.2299792 PM	WindowsTerminal.exe	77076	Process Create	C:\WINDOWS\system32\wsl.exe	SUCCESS	PID: 78492, Command line: "wsl.exe" --cd "C:/etc" -d Debian
```

**Review by zadjii-msft — APPROVED** [review]

**Comment by LuanVSO** [comment]
this change makes osc 9;9 work with linux paths but breaks `\\wsl$` ones 
docs at https://docs.microsoft.com/windows/terminal/tutorials/new-tab-same-directory will need to be updated

**Pull request merged**